File: Sources/App/Classes/Views/Main Window/TVCMainWindowTextViewAppearance.m
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\\__|\__,_|\__,_|_|
 *
 *    Copyright (c) 2018 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

#import "NSObjectHelperPrivate.h"
#import "TVCMainWindow.h"
#import "TVCAppearancePrivate.h"
#import "TVCMainWindowTextViewAppearancePrivate.h"

NS_ASSUME_NONNULL_BEGIN

@interface TVCMainWindowTextViewAppearance ()
#pragma mark -
#pragma mark Text View

@property(nonatomic, assign, readwrite) NSSize textViewInset;
@property(nonatomic, copy, nullable, readwrite) NSColor *textViewTextColor;
@property(nonatomic, copy, nullable, readwrite) NSColor *textViewPlaceholderTextColor;

@property(nonatomic, assign, readwrite) TVCMainWindowTextViewFontSize textViewPreferredFontSize;

#pragma mark -
#pragma mark Background View

@property(nonatomic, assign, readwrite) CGFloat backgroundViewContentBorderPadding;
@end

@implementation TVCMainWindowTextViewAppearance

#pragma mark -
#pragma mark Initialization

- (nullable instancetype)initWithWindow:(TVCMainWindow *)mainWindow
{
	NSParameterAssert(mainWindow != nil);

	NSURL *appearanceLocation = [self.class appearanceLocation];

	BOOL forRetinaDisplay = mainWindow.runningInHighResolutionMode;

	if ((self = [super initWithAppearanceAtURL:appearanceLocation forRetinaDisplay:forRetinaDisplay])) {
		[self prepareInitialState];

		return self;
	}

	return nil;
}

+ (NSURL *)appearanceLocation
{
	return [RZMainBundle() URLForResource:@"TVCMainWindowTextViewAppearance" withExtension:@"plist"];
}

- (void)prepareInitialState
{
	NSDictionary *properties = self.appearanceProperties;

	NSDictionary *textView = properties[@"Text View"];

	self.textViewInset = [self sizeInGroup:textView withKey:@"inset"];
	self.textViewTextColor = [self colorInGroup:textView withKey:@"normalTextColor"];
	self.textViewPlaceholderTextColor = [self colorInGroup:textView withKey:@"placeholderTextColor"];

	NSDictionary *backgroundView = properties[@"Background View"];

	self.backgroundViewContentBorderPadding = [self measurementInGroup:backgroundView withKey:@"contentBorderPadding"];

	[self flushAppearanceProperties];
}

#pragma mark -
#pragma mark Everything Else

- (BOOL)preferredTextViewFontChanged
{
	return (self.textViewPreferredFontSize != [TPCPreferences mainTextViewFontSize]);
}

- (nullable NSFont *)textViewPreferredFont
{
	TVCMainWindowTextViewFontSize preferredFontSize = [TPCPreferences mainTextViewFontSize];

	self.textViewPreferredFontSize = preferredFontSize;

	/* Sizes track the system text styles so they follow the
	 user's text size preferences rather than fixed point values. */
	switch (preferredFontSize) {
	case TVCMainWindowTextViewFontSizeLarge:
		return [NSFont preferredFontForTextStyle:NSFontTextStyleTitle3 options:@{}];
	case TVCMainWindowTextViewFontSizeExtraLarge:
		return [NSFont preferredFontForTextStyle:NSFontTextStyleTitle2 options:@{}];
	case TVCMainWindowTextViewFontSizeHumongous:
		return [NSFont preferredFontForTextStyle:NSFontTextStyleTitle1 options:@{}];
	default:
		return [NSFont preferredFontForTextStyle:NSFontTextStyleBody options:@{}];
	}
}

@end

NS_ASSUME_NONNULL_END

File: Sources/App/Modules/AppKitSupport/TVCMainWindowTextViewAppearance.swift
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \\ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\\___/_/\\_\\__|\\__,_|\\__,_|_|
 *
 * Copyright (c) 2010 - 2026 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 *********************************************************************** */

import AppKit

@objc(TVCMainWindowTextViewAppearance)
public final class MainWindowTextViewAppearance: ApplicationAppearance {
	@objc public private(set) var textViewInset: NSSize = .zero
	@objc public private(set) var textViewTextColor: NSColor?
	@objc public private(set) var textViewPlaceholderTextColor: NSColor?
	@objc public private(set) var textViewPreferredFontSize: TVCMainWindowTextViewFontSize = .normal
	@objc public private(set) var backgroundViewContentBorderPadding: CGFloat = 0

	@objc(initWithWindow:)
	public init?(window _: TVCMainWindow) {
		guard let appearanceLocation = Bundle.main.url(
			forResource: "TVCMainWindowTextViewAppearance",
			withExtension: "plist"
		) else {
			return nil
		}

		super.init(appearanceAt: appearanceLocation)

		guard let properties = appearanceProperties else {
			return nil
		}

		let textView = properties["Text View"] as? [String: Any] ?? [:]
		textViewInset = size(inGroup: textView, withKey: "inset")
		textViewTextColor = color(inGroup: textView, withKey: "normalTextColor")
		textViewPlaceholderTextColor = color(inGroup: textView, withKey: "placeholderTextColor")

		let backgroundView = properties["Background View"] as? [String: Any] ?? [:]
		backgroundViewContentBorderPadding = measurement(inGroup: backgroundView, withKey: "contentBorderPadding")

		flushAppearanceProperties()
	}

	@objc public func preferredTextViewFontChanged() -> Bool {
		textViewPreferredFontSize != TPCPreferences.mainTextViewFontSize()
	}

	@objc public var textViewPreferredFont: NSFont? {
		let preferredFontSize = TPCPreferences.mainTextViewFontSize()
		textViewPreferredFontSize = preferredFontSize

		/* Sizes track the system text styles so they follow the
		 user's text size preferences rather than fixed point values. */
		switch preferredFontSize {
		case .large:
			return NSFont.preferredFont(forTextStyle: .title3, options: [:])
		case .extraLarge:
			return NSFont.preferredFont(forTextStyle: .title2, options: [:])
		case .humongous:
			return NSFont.preferredFont(forTextStyle: .title1, options: [:])
		default:
			return NSFont.preferredFont(forTextStyle: .body, options: [:])
		}
	}
}
